pmm/queries.go

```diff
@@ -0,0 +1,149 @@
+/*
+	Copyright (c) 2016, Percona LLC and/or its affiliates. All rights reserved.
+
+	This program is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with this program.  If not, see <http://www.gnu.org/licenses/>
+*/
+
+package pmm
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"time"
+
+	"github.com/percona/pmm/proto"
+	protocfg "github.com/percona/pmm/proto/config"
+)
+
+// deleteInstance delete instance on QAN API.
+func (a *Admin) deleteInstance(uuid string) error {
+	// Remove MySQL instance from QAN.
+	url := a.qanAPI.URL(a.serverURL, qanAPIBasePath, "instances", uuid)
+	resp, content, err := a.qanAPI.Delete(url)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != http.StatusNoContent {
+		return a.qanAPI.Error("DELETE", url, resp.StatusCode, http.StatusNoContent, content)
+	}
+
+	return nil
+}
+
+// getAgentInstance get agent instance from QAN API and return its parent_uuid.
+func (a *Admin) getAgentInstance(agentID string) (string, error) {
+	var in proto.Instance
+	url := a.qanAPI.URL(a.serverURL, qanAPIBasePath, "instances", agentID)
+	resp, bytes, err := a.qanAPI.Get(url)
+	if err != nil {
+		return "", err
+	}
+	if resp.StatusCode == http.StatusNotFound {
+		// No agent instance on QAN API - orphaned agent installation.
+		return "", errNoInstance
+	}
+	if resp.StatusCode != http.StatusOK {
+		return "", a.qanAPI.Error("GET", url, resp.StatusCode, http.StatusOK, bytes)
+	}
+
+	if err := json.Unmarshal(bytes, &in); err != nil {
+		return "", err
+	}
+
+	return in.ParentUUID, nil
+}
+
+// getAgentID read agent UUID from agent QAN config file.
+func getAgentID(configFile string) (string, error) {
+	jsonData, err := ioutil.ReadFile(configFile)
+	if err != nil {
+		return "", err
+	}
+
+	config := &protocfg.Agent{}
+	if err := json.Unmarshal(jsonData, &config); err != nil {
+		return "", err
+	}
+
+	return config.UUID, nil
+}
+
+// manageQAN enable/disable QAN on agent through QAN API.
+func (a *Admin) manageQAN(agentID, cmdName, UUID string, config map[string]interface{}) error {
+	var data []byte
+	if cmdName == "StartTool" {
+		data, _ = json.Marshal(config)
+	} else if cmdName == "StopTool" {
```

Comment:
Yes, but `pmm-admin` doesn't use that

On installing service it sends cmd `StartTool` and on removing service it sends cmd `StopTool`.

To make this more clear and proper I've made changes:
* don't depend on string param that needs to be either StartTool or StopTool and if it's not StartTool or StopTool it brakes
* introduce `startQAN(agentID, qanConfig)` and `stopQAN(agentID, UUID)` that do what they say with different set of args; previously we had `manageQAN(agentID, "StartTool", "", qanConfig)` and `manageQAN(agentID, "StopTool", UUID, nil)` so you needed to know that with "StartTool" string you need to provide `qanConfig` but no `uuid` and reverse for "StopTool".